qiskit_aer/backends/plugin/aer_backend_plugin.py

```diff
@@ -0,0 +1,116 @@
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Aer simulator backend transpiler plug-in
+"""
+from qiskit.transpiler.preset_passmanagers.plugin import PassManagerStagePlugin
+from qiskit.transpiler import PassManager, TransformationPass
+from qiskit.transpiler.passes import BasisTranslator
+from qiskit.transpiler.passes import UnitarySynthesis
+from qiskit.transpiler.passes import HighLevelSynthesis
+from qiskit.circuit.equivalence_library import SessionEquivalenceLibrary as sel
+from qiskit.circuit.measure import Measure
+from qiskit.circuit import ControlFlowOp
+from qiskit.converters import circuit_to_dag
+from qiskit_aer.backends.name_mapping import NAME_MAPPING
+
+
+class AerBackendRebuildGateSetsFromCircuit(TransformationPass):
+    """custom translation class to rebuild basis gates with gates in circuit"""
+
+    def __init__(self, config, opt_lvl):
+        super().__init__()
+        self.config = config
+        if opt_lvl is None:
+            self.optimization_level = 1
+        else:
+            self.optimization_level = opt_lvl
+
+    def _add_ops(self, dag, ops):
+        num_unsupported_ops = 0
+        opnodes = dag.op_nodes()
+        if opnodes is None:
+            return num_unsupported_ops
+
+        for node in opnodes:
+            if isinstance(node.op, ControlFlowOp):
+                for block in node.op.blocks:
+                    num_unsupported_ops = num_unsupported_ops + self._add_ops(
+                        circuit_to_dag(block), ops
+                    )
+            if node.name in self.config.target:
+                if node.name not in ops:
+                    ops.append(node.name)
+            else:
+                num_unsupported_ops = num_unsupported_ops + 1
+        return num_unsupported_ops
+
+    def run(self, dag):
+        # do nothing for higher optimization level
+        if self.optimization_level > 1:
+            return dag
+        if self.config is None or self.config.target is None:
+            return dag
+
+        # search ops in supported name mapping
+        ops = []
+        num_unsupported_ops = self._add_ops(dag, ops)
+
+        # if there are some unsupported node (i.e. RealAmplitudes) do nothing
+        if num_unsupported_ops > 0 or len(ops) < 1:
+            return dag
+
+        # clear all instructions in target
+        self.config.target._gate_map.clear()
+        self.config.target._gate_name_map.clear()
+        self.config.target._qarg_gate_map.clear()
+        self.config.target._global_operations.clear()
+
+        # rebuild gate sets from circuit
+        for name in ops:
+            if name not in self.config.target:
+                if name != "measure":
+                    self.config.target.add_instruction(NAME_MAPPING[name], name=name)
```

Comment:
I'm not sure this will do the correct thing in the case of a noise model. If you have a noise model won't this remove all the connectivity constraints caused by that? (this also connects with my comment below, you'd need to update the coupling map for any optimization or scheduling plugins).